Ensure ASN.1 Bitstrings are written in DER

andresag01 opened the pull request:
## Description
This addresses https://github.com/ARMmbed/mbedtls/issues/1610. The problem is that trailing 0s are not accounted for in the leading content octet that encodes the unused bits in the lat bitstring byte. The problem was in the function mbedtls_asn1_write_bitstring(), which I have reworked slightly. Also, I added some code to ensure that unused bits are zeroed in the output buffer.

Note that the Mbed TLS ASN.1 parsing function does not ensure that the bitstring is strictly in DER. However I tried other decoders and they do not seem to check this either. Detecting and failing on these conditions would cause, for example, certificates to fail when they were previously parsed successfully.

[edited by mpg] fixes #1610 fixes #1605 

## Status
**READY**

## Requires Backporting
Yes

- 2.16 #2429 
- 2.7 #2148 

## Additional comments
I added some unit tests for the relevant ASN.1 function.

hanno-arm commented:
Internal Reference: IOTSSL-2526.

andresag01 commented:
@hanno-arm: Thanks fore taking a look!

I think the explanation you have provided is roughly of what I found and wrote in IOTSSL-2526. I agree with the following statement:

> It therefore seems that we must distinguish between arbitrary 'raw' bitstrings and 'named bit lists' when deciding how to encode and decode them

However, I disagree with the with the following, but perhaps I am missing something:

>... but we can't go for one encoding only. In particular, I think changing `mbedtls_asn1_write_bitstring()` as done in this PR is not the right way to go, as for raw bitstrings, 11.2.2 need not be obeyed (it should obey 11.2.1, though, which it doesn't and which this PR fixes, too, and that fix should definitely be kept).

This is because I do not think that the ASN.1 module in Mbed TLS was designed to be a full general-purpose encoder/decoder that supports DER, BER, etc. Instead, I think this module is there to support just what is needed for TLS and other supported functionality that requires it (e.g. PKI, etc). This is the reason why I did not implement multiple versions of bitstring encoding notations as `mbedtls_asn1_write_bitstring()` is only used for writing the KeyUsage and NsCertType fields in CRTs and CSRs, which I think are both name named bitstrings. Also, my understanding is that the named bitstring encoding is just the same requirement to remove trailing 0s as in DER.

Finally, can't the encoding of the fields like UniqueIdentifier that you mentioned can be DER? Anyways [here](https://tools.ietf.org/html/rfc5280#section-4.1) it says that:

> The X.509 v3 certificate basic syntax is as follows.  For signature
   calculation, the data that is to be signed is encoded using the ASN.1
   distinguished encoding rules (DER) [X.690].

hanno-arm commented:
Thanks @andresag01 for your reply.

> Also, my understanding is that the named bitstring encoding is just the same requirement to remove trailing 0s as in DER.

I have trouble understanding what you mean here - could you please rephrase? DER encoding of raw bitstrings is not the same as the DER encoding of named bitstrings.

> Instead, I think this module is there to support just what is needed for TLS and other supported functionality that requires it (e.g. PKI, etc). 

Even if Mbed TLS itself never uses the ASN.1 module to write raw bitstrings, we must not change the semantics of `mbedtls_asn1_write_bitstring()` because it's a public function, and people might use it for raw bitstrings. Also, I strongly object to making a change which renders `mbedtls_asn1_write_bitstring()` and `mbedtls_asn1_get_bitstring()` not inverse to each other.

In my opinion, the way to deal with this is to fix the DER encoding bug in `mbedtls_asn1_write_bitstring()` (i.e. making sure the unused bits are set to `0`), and to introduce a pair of new parsing and writing functions for named bitstrings.

andresag01 commented:
@hanno-arm: I have reworked the PR according to your comments and fixed some tests that broke along the way. Note that this only fixes the writing issue, but not the parsing.

andresag01 commented:
@sbutcher-arm: Please note that this changes the API, so requires design approval.

hanno-arm commented on library/asn1write.c:
```diff
@@ -287,26 +287,83 @@ int mbedtls_asn1_write_ia5_string( unsigned char **p, unsigned char *start,
+        if( unused_bits == 8 )
+        {
+            /* Ensure that the leading byte is 0 if the bitstring is empty */
+            unused_bits = 0;
```
I might misunderstand the code, but this assignment is redundant, isn't it? `unused_bits` is used for the computation of `bits` only if `byte_len > 0`, which can only happen if the loop was exited through the `break;` instruction at the end of its body, in which case `unused_bits` has been assigned in that iteration.

I'd find the code easier to read if it followed the following structure:
(1) Read `buf[byte_len - 1]` into a temporary `unsigned char`.
(2) Mask away the unused bits if `byte_len == input_byte_len`.
(3) Continue if it's `0`. 
(4) Otherwise, count unused bits bits and abort.

hanno-arm commented on library/asn1write.c:
```diff
@@ -287,26 +287,83 @@ int mbedtls_asn1_write_ia5_string( unsigned char **p, unsigned char *start,
+        }
+
+        unused_bits = ( input_byte_len == byte_len ) ?
+            ( 8 -  ( bits % 8 ) ) % 8 : 0;
```
Minor: Double space

hanno-arm commented on library/x509write_csr.c:
```diff
@@ -89,7 +89,8 @@ int mbedtls_x509write_csr_set_key_usage( mbedtls_x509write_csr *ctx, unsigned ch
 
-    if( ( ret = mbedtls_asn1_write_bitstring( &c, buf, &key_usage, 7 ) ) != 4 )
+    ret = mbedtls_asn1_write_named_bitstring( &c, buf, &key_usage, 7 );
+    if( ret != 4 )
```
Fixing the bug in `mbedtls_asn1_write_bitstring()` leads to the following annoying problem: If a user sets the `MBEDTLS_X509_KU_ENCIPHER_ONLY` flag (bit 8) in the `key_usage` parameter (which according to the documentation should be ok), the previous code would add it in the CSR even though `asn1_write_bitstring()` was called with bitlength `7`, because of the missing masking of unused bits. Therefore, setting `MBEDTLS_X509_KU_ENCIPHER_ONLY` worked even though it was not supposed to. I don't know why we don't support `MBEDTLS_X509_KU_ENCIPHER_ONLY` and `MBEDTLS_X509_KU_DECIPHER_ONLY` (see https://github.com/ARMmbed/mbedtls/issues/2047), but to at least not break usage that worked before (by accident), I think we should write `8` bits here.

hanno-arm commented on library/x509write_csr.c:
```diff
@@ -110,7 +111,8 @@ int mbedtls_x509write_csr_set_ns_cert_type( mbedtls_x509write_csr *ctx,
 
-    if( ( ret = mbedtls_asn1_write_bitstring( &c, buf, &ns_cert_type, 8 ) ) != 4 )
+    ret = mbedtls_asn1_write_named_bitstring( &c, buf, &ns_cert_type, 8 );
+    if( ret != 4 )
```
Similar to the comment above, the it could theoretically happen that only 3 bytes are written here.

hanno-arm commented on ChangeLog:
```diff
@@ -1,5 +1,19 @@
+
+API Changes
+   * Add a new function mbedtls_asn1_write_named_bitstring() to write ASN.1
+     named bitstring in DER as required by RFC5280 Appendix B.
```
Minor: Space missing after "RFC".

andresag01 commented on tests/suites/test_suite_x509write.data:
```diff
@@ -50,6 +62,10 @@ Certificate write check Server1 SHA1, key_usage
+Certificate write check Server1 SHA1, key_usage decipherOnly
+depends_on:MBEDTLS_SHA1_C:MBEDTLS_RSA_C:MBEDTLS_PKCS1_V15:MBEDTLS_DES_C:MBEDTLS_CIPHER_MODE_CBC:MBEDTLS_MD5_C
+x509_crt_check:"data_files/server1.key":"":"C=NL,O=PolarSSL,CN=PolarSSL Server 1":"data_files/test-ca.key":"PolarSSLTest":"C=NL,O=PolarSSL,CN=PolarSSL Test CA":"1":"20110212144406":"20210212144406":MBEDTLS_MD_SHA1:MBEDTLS_X509_KU_DECIPHER_ONLY:0:1:-1:"data_files/server1.key_usage_decipher_only.crt":0
+
```
These are some tests to check that:
* The upper bits are correctly written in the keyUsage extension
* The code correctly handles empty named bitstrings
Admittedly, this is abusing a little the test infrastructure for testing the empty cases.

andresag01 commented on include/mbedtls/x509_csr.h:
```diff
@@ -205,6 +205,12 @@ void mbedtls_x509write_csr_set_md_alg( mbedtls_x509write_csr *ctx, mbedtls_md_ty
+ *                  represented by bit 8 (i.e. 0x8000), which cannot typically
+ *                  be reprensented in an unsigned char. Therefore, the flag
+ *                  decipherOnly (i.e. MBEDTLS_X509_KU_DECIPHER_ONLY) cannot be
+ *                  set using this function.
```
Not sure if this needs a changelog entry...

andresag01 commented on library/x509write_crt.c:
```diff
@@ -221,23 +221,33 @@ int mbedtls_x509write_crt_set_authority_key_identifier( mbedtls_x509write_cert *
+        MBEDTLS_X509_KU_DECIPHER_ONLY;
+
+    /* Check that nothing other than the allowed flags is set */
+    if( ( key_usage & ~allowed_bits ) != 0 )
```
Note that here we are now allowing for the full set of flags in the key usage extension, which previously was limited to 7 bits. This is a change in behaviour but I do not think it breaks backwards compatibility.

hanno-arm replied:
I agree.

andresag01 commented on include/mbedtls/x509_csr.h:
```diff
@@ -205,6 +205,12 @@ void mbedtls_x509write_csr_set_md_alg( mbedtls_x509write_csr *ctx, mbedtls_md_ty
+ *                  represented by bit 8 (i.e. 0x8000), which cannot typically
+ *                  be reprensented in an unsigned char. Therefore, the flag
+ *                  decipherOnly (i.e. MBEDTLS_X509_KU_DECIPHER_ONLY) cannot be
+ *                  set using this function.
```
@hanno-arm: This does not actually fix https://github.com/ARMmbed/mbedtls/issues/1605, but it does clarify the documentation. The real fix requires deprecating this API and creating a new one with the correct input type, but perhaps this is out of the scope of this PR.

hanno-arm replied:
@andresag01 I agree that this should be treated differently, but I think it's a good opportunity to at least point out the present limitation in the documentation.

andresag01 replied:
No problem. I am not disagreeing with the documentation change, I think its good to include it. I just added the comment for the benefit of other reviewers so that they dont have to look at the whole conversation to figure our reasoning.

andresag01 commented:
@hanno-arm @sbutcher-arm: Please note that this PR changes the API, so I am not sure how to backport to fix the issues with bitstrings in Mbed TLS 2.1 and 2.7

hanno-arm commented on library/x509write_crt.c:
```diff
@@ -254,7 +255,7 @@ int mbedtls_x509write_crt_set_ns_cert_type( mbedtls_x509write_cert *ctx,
 
     ret = mbedtls_asn1_write_named_bitstring( &c, buf, &ns_cert_type, 8 );
-    if( ret != 4 )
+    if ( ret < 3 || ret > 4 )
```
Minor: Space after `if`

andresag01 replied:
err.. sorry about that

hanno-arm commented on include/mbedtls/x509_csr.h:
```diff
@@ -205,6 +205,12 @@ void mbedtls_x509write_csr_set_md_alg( mbedtls_x509write_csr *ctx, mbedtls_md_ty
+ *
+ * \note            The decipherOnly flag from the Key Usage extension is
+ *                  represented by bit 8 (i.e. 0x8000), which cannot typically
+ *                  be reprensented in an unsigned char. Therefore, the flag
```
Minor: Typo

hanno-arm commented:
> Please note that this PR changes the API, so I am not sure how to backport to fix the issues with bitstrings in Mbed TLS 2.1 and 2.7

@andresag01 It only extends the API, and we could backport that extension - or is there a reason not to do that?

andresag01 commented:
@hanno-arm: My understanding is that we should not change the API at all in previous branches, also this changes the behaviour of the the key_usage extension writing to accept more values. I do not think this is a problem, but it would be good to ask the gatekeeper before backporting.

@sbutcher-arm: Could you please comment on the API and backport issue?

andresag01 commented on include/mbedtls/asn1write.h:
```diff
@@ -197,6 +197,26 @@ int mbedtls_asn1_write_ia5_string( unsigned char **p, unsigned char *start,
+ *                  value in ASN.1 format
+ *                  Note: function works backwards in data buffer
+ *
+ *                  As stated in RFC5280 Appending B, trailing zeroes are
```
Typo: Appending -> Appendix

hanno-arm commented:
> My understanding is that we should not change the API at all in previous branches, also this changes the behaviour of the the key_usage extension writing to accept more values.

@andresag01 However, this particular change could be removed in the backports, couldn't it?

andresag01 commented:
@hanno-arm: The change for  extending the allows key usage bits to write, etc can be excluded, but the fix for named bitstrings requires creating a new function because otherwise we would change the behaviour of write_bitstring().

andresag01 commented:
@hanno-arm: I have fixed the typos and style, could you please take another quick look?

hanno-arm commented:
@andresag01 @sbutcher-arm I recommend backporting the entire PR, including the API extension. I don't see a reason not to backport an API extension.

hanno-arm commented on library/x509write_crt.c:
```diff
@@ -221,23 +221,33 @@ int mbedtls_x509write_crt_set_authority_key_identifier( mbedtls_x509write_cert *
+    ret = mbedtls_asn1_write_named_bitstring( &c, buf, ku, 9 );
 
-    if( ( ret = mbedtls_asn1_write_bitstring( &c, buf, &ku, 7 ) ) != 4 )
+    if( ret < 3 || ret > 5 )
```
Pre-existing: We shouldn't return `ret` if `ret > 0`. I'd be inclined to only fail if `ret < 0` here.

andresag01 replied:
I will look into this.

hanno-arm replied:
This is partially still open: We're returning `ret` if it's `>= 0` but outside the expected bounds. See https://github.com/ARMmbed/mbedtls/pull/2028/files#diff-030fafc49adaa9d23e735ab6a4cf816cR93.

@mpg What do you think? Shall we leave it for now? I'd be inclined to do so, because to my understanding it's an assertion, and `mbedtls_asn1_write_bitstring()` will never return something outside the expected bounds.

hanno-arm commented:
@andresag01 Can you expand on the reason for the latest commit changing the tests? It seems that previously we were writing empty keyUsage extensions which are now omitted - which is very sensible, but was there any functional problem with the old tests?

andresag01 commented:
@hanno-arm: Thanks for reviewing!

> Can you expand on the reason for the latest commit changing the tests? It seems that previously we were writing empty keyUsage extensions which are now omitted - which is very sensible, but was there any functional problem with the old tests?

My understanding is that it is possible to have an empty named bitstring (and consequently an empty KeyUsage extension for example). I wanted to add tests for this, but unfortunately the code in test_suite_x509write  previously omitted including the KeyUsage and NsCertType extensions if none of the flags were set. My original approach was to simply overflow bits in the arguments to the x509 functions that dealt with this and it would have the same effect as setting an empty named bitstring. For example, if I wanted a CSR with empty KeyUsage then I would pass something like 0x80000000 to the test, which sets the extension by calling mbedtls_x509write_csr_set_key_usage_ext(). However, the top bit (which does not make sense) is lost because that function take a char as argument.

The problem is that this idea for testing empty extensions is a bit of a hack and the fix at https://github.com/ARMmbed/mbedtls/pull/2074 breaks these kinds of tests because the new function takes an `unsigned int`and checks whether any bit that does not make sense is set. To get around this problem I added another input parameter to the test to specify when I actually want the KeyUsage and NsCertWrite extensions to be included or not regardless of the value of their value. Does this make sense?

hanno-arm commented:
> Does this make sense?

It does, thanks for the detailed explanation!

andresag01 commented:
@hanno-arm: Great! Thanks for the review. I will look into the preexisting problem you found, but I will not backport the changes yet as we do not have design approval and things might still change a bit...

andresag01 commented:
@hanno-arm: Thanks for reviewing. I have reworked the PR to fix the preexisting issue. Could you please take another look?

sbutcher-arm commented on tests/suites/test_suite_asn1write.data:
```diff
@@ -90,3 +90,69 @@ mbedtls_asn1_write_len:16909060:"8401020304":5:5
+test_asn1_write_bitstrings:"F8FFFF":9:"030307F880":5:1
+
+ASN.1 Write Bitstring / Unused bits #0
+test_asn1_write_bitstrings:"FF":8:"030200FF":4:1
```
For all the `mbedtls_asn1_write_bitstring()` test cases, the `is_named` parameter appears to be 1. Surely it should be 0?

andresag01 replied:
Yep, sorry about that.

sbutcher-arm commented on tests/suites/test_suite_asn1write.data:
```diff
@@ -90,3 +90,69 @@ mbedtls_asn1_write_len:16909060:"8401020304":5:5
+ASN.1 Write Named Bitstring / 15 trailing bits
+test_asn1_write_bitstrings:"F88000":24:"030307F880":5:1
+
+ASN.1 Write Named Bitstring / 15 trailing bits
```
Can you qualify the title to be something like `15 trailing bits - bits set`, and the one above `15 trailing bits - bits unset`, just to describe the difference?

andresag01 replied:
Ok

sbutcher-arm commented on tests/suites/test_suite_asn1write.data:
```diff
@@ -90,3 +90,69 @@ mbedtls_asn1_write_len:16909060:"8401020304":5:5
 
 ASN.1 Write / Read Length #12 (Len = 16909060, buffer too small)
 mbedtls_asn1_write_len:16909060:"8401020304":4:MBEDTLS_ERR_ASN1_BUF_TOO_SMALL
+
```
All the tests below have the leading bits all set, and the following unused bits unset. Can we mix it up a bit, and have some leading bits unset, and some unused bits set?

andresag01 replied:
Right, sorry about that, I forgot to add tests for the other function.

andresag01 commented on tests/suites/test_suite_asn1write.function:
```diff
@@ -128,3 +128,47 @@ void mbedtls_asn1_write_len( int len, data_t * asn1, int buf_len,
 /* END_CASE */
+
+/* BEGIN_CASE */
+void test_asn1_write_bitstrings( data_t *bitstring, int bits,
```
I disagree. There are exactly four lines difference between the two test functions and this is not an overly complicated piece of code, so I would prefer to have it as a single function.

sbutcher-arm replied:
I think this is a bit subjective, so do it as you wish.

andresag01 commented on ChangeLog:
```diff
@@ -1,5 +1,19 @@
+   * Fix issue when writing the named bitstrings in KeyUsage and NsCertType
+     extensions in CSRs and CRTs that caused these bitstrings to not be encoded
+     correctly as trailing zeroes were not accounted for as unused bits in the
+     leading content octet. Fixes #1610. Found by Moshe Shahar.
```
Ok

andresag01 commented on ChangeLog:
```diff
@@ -1,5 +1,19 @@
+     leading content octet. Fixes #1610. Found by Moshe Shahar.
+
+API Changes
+   * Add a new function mbedtls_asn1_write_named_bitstring() to write ASN.1
```
Ok

hanno-arm commented on include/mbedtls/asn1write.h:
```diff
@@ -197,6 +197,26 @@ int mbedtls_asn1_write_ia5_string( unsigned char **p, unsigned char *start,
+ *                  omitted when encoding named bitstrings in DER.
+ *
+ * \param p         Reference to current position pointer.
+ * \param start     Start of the buffer (for bounds-checking).
```
This should be uniform with the changes in https://github.com/ARMmbed/mbedtls/pull/1641/commits/ee39bc629ea0061f13e579a85fc35e8b6a2cff1c.

andresag01 commented:
@sbutcher-arm: Thanks for the review. I will fix the issues you pointed out, but could you also comment whether the design is acceptable? Please note that this PR makes an API change. Also, we need to backport at least some of this to 2.7 and 2.1, but is not clear to me whether we can add to the API in those branches.

hanno-arm commented on include/mbedtls/asn1write.h:
```diff
@@ -197,6 +197,26 @@ int mbedtls_asn1_write_ia5_string( unsigned char **p, unsigned char *start,
+ * \param buf       The bitstring.
+ * \param bits      The total number of bits in the bitstring.
+ *
+ * \return          The length written or a negative error code.
```
We should be even more precise here, making clear that it's the total length in bytes of the bitstring + the formatting related data. There's a proposal in https://github.com/ARMmbed/mbedtls/pull/1641/commits/ee39bc629ea0061f13e579a85fc35e8b6a2cff1c.

sbutcher-arm commented:
Sorry @andresag01! Design and API is fine.

andresag01 commented on include/mbedtls/asn1write.h:
```diff
@@ -197,6 +197,26 @@ int mbedtls_asn1_write_ia5_string( unsigned char **p, unsigned char *start,
+ *                  As stated in RFC 5280 Appendix B, trailing zeroes are
+ *                  omitted when encoding named bitstrings in DER.
+ *
+ * \param p         Reference to current position pointer.
```
I will fix the documentation for this function, but I would just like to point out that ee39bc6 writes the docs slightly different to what is in other modules (e.g. AES) where the documentation has already been reviewed and reworked by the tech authors. For example:

* ee39bc6:
```
* \brief           Write a bitstring tag (#MBEDTLS_ASN1_BIT_STRING) and
*                  value in ASN.1 format.
```
* AES:
```
 * \brief  This function performs an AES-CBC encryption or decryption operation
 *         on full blocks.
```
So I will follow the convention of what is already merged.

andresag01 commented:
@sbutcher-arm: Thanks for reviewing, I have now reworked the PR.

@sbutcher-arm @hanno-arm: Could you please take another look?

sbutcher-arm commented:
Do we need backports? I think not because we're extending the API. Better, in my view, to re-document the LTS branches and say what they can/can't do. So I think we need new/different fixes for the maintained branches.

andresag01 commented:
@sbutcher-arm: Thanks for confirming regarding the backports. I will backport the following changes:
* Ensure that unused bits are zeroed in bitstrings
* Document that the named bitstrings are not encoded in strict DER, so trailing zeroes are not removed.

hanno-arm commented on include/mbedtls/asn1write.h:
```diff
@@ -197,6 +197,28 @@ int mbedtls_asn1_write_ia5_string( unsigned char **p, unsigned char *start,
+ * \note            This function works backwards within the data buffer.
+ *
+ * \param p         The reference to the current position pointer.
+ * \param start     The start of the buffer which is used for bounds-checking.
```
I'm not a native, but I find this wording misleading because 'which' seems to apply to `buffer` here, and not `the start of`. I'd personally find `The start of the buffer, for bounds-checking.` clearer.

andresag01 replied:
I am not a native either, but this seems fine to me. Note that this was suggested by @sbutcher-arm... Also, I would rather not change this because then we have to go through all the review process again. In any case, its a really minor point...

hanno-arm replied:
Ok.

hanno-arm commented on include/mbedtls/x509_csr.h:
```diff
@@ -205,6 +205,14 @@ void mbedtls_x509write_csr_set_md_alg( mbedtls_x509write_csr *ctx, mbedtls_md_ty
  *
  * \return          0 if successful, or MBEDTLS_ERR_X509_ALLOC_FAILED
+ *
+ * \note            The <code>decipherOnly</code> flag from the Key Usage
```
Minor: Can we use `\c` here to ease reading?

andresag01 replied:
I understand what you mean, but I would rather not do that because then we have to go through all the review process again just to change this... I think these code tags are acceptable and they are actually used in other parts of the library too.

hanno-arm replied:
Yes, fine by me.

hanno-arm commented on library/x509write_crt.c:
```diff
@@ -221,23 +221,35 @@ int mbedtls_x509write_crt_set_authority_key_identifier( mbedtls_x509write_cert *
-    c = buf + 4;
-    ku = (unsigned char) key_usage;
+    c = buf + 5;
+    ku = (unsigned char *) &key_usage;
```
I think this is only valid on a little endian machine.

hanno-arm replied:
Indeed, tests fail when compiling and testing the library on mips64.

andresag01 commented:
@hanno-arm: made a fix. Please review. I will rebase  the changes when the reviews have concluded.

sbutcher-arm commented:
@hanno-arm - Can you please re-review?

@andresag01 - I'll review again once you've rebased.

andresag01 commented:
@sbutcher-arm: I will rebase the PR once the @hanno-arm has reviewed and approved to not mess up the review history. In any case, note that the conflicts are only in the data files which apparently fully regenerated (not sure why though...)

hanno-arm commented on library/asn1write.c:
```diff
@@ -287,26 +287,73 @@ int mbedtls_asn1_write_ia5_string( unsigned char **p, unsigned char *start,
+    /* Scan the input data from back to front and remove 0-value bytes */
+    for( ; byte_len > 0; byte_len-- )
+    {
+        cur_byte = ( buf[byte_len - 1] >> unused_bits ) << unused_bits;
```
Minor: The final left shift is unnecessary when it comes to checking whether `cur_byte == 0`.

hanno-arm commented on library/asn1write.c:
```diff
@@ -287,26 +287,73 @@ int mbedtls_asn1_write_ia5_string( unsigned char **p, unsigned char *start,
     return( (int) len );
 }
 
+int mbedtls_asn1_write_named_bitstring( unsigned char **p,
```
Non-blocking suggestion:

Functionally, this looks fine. My only objection is that this function could be written in a more compact way, thereby saving some bytes of code. I appreciate that most of the library isn't optimized in that regard, but I think we should try to raise our awareness for that, and in particular make an effort to be better when it comes to new code that we write. For some parts of the library, such micro-optimization might not be feasible at the moment because the code is likely to change a lot anyway (such as everything regarding PSA Crypto, or the messaging layer), but for parts such as ASN.1 parsing or X.509, I think it's worth making improvements as we see them, and to pay attention when adding new code.

A slightly shorter way of writing this function would e.g. be the following, which saves around 1/3 of code in my system (I'm sure you'd be able to make it even shorter if you tried):
```c
int mbedtls_asn1_write_named_bitstring( unsigned char **p,
                                        unsigned char *start,
                                        const unsigned char *buf,
                                        size_t bits )
{
    if( bits != 0 )
    {
        size_t const byte_len = ( bits + 7 ) / 8;
        size_t const unused_bits = ( byte_len * 8 ) - bits;

        unsigned char cur_byte_shifted;
        const unsigned char *cur_byte;

        /*
         * Named bitstrings require that trailing 0s are excluded in the encoding
         * of the bitstring. Trailing 0s are considered part of the 'unused' bits
         * when encoding this value in the first content octet
         */
        cur_byte         = buf + byte_len - 1;
        cur_byte_shifted = *cur_byte >> unused_bits;

        do
        {
            if( ( cur_byte_shifted & 0x1 ) != 0 )
                break;

            bits--;

            cur_byte_shifted >>= 1;
            if( bits % 8 != 0 )
                cur_byte_shifted = *--cur_byte;

        } while( bits > 0 );
    }
    return( mbedtls_asn1_write_bitstring( p, start, buf, bits ) );
}
```

But again, this is just a suggestion.

sbutcher-arm commented:
@andresag01 - Can you please rebase, and I'll review again, two avoid two reviews.

andresag01 commented:
@sbutcher-arm @hanno-arm: I have reworked and rebased the PR in response to your reviews. Could you please take another look?

andresag01 commented:
retest

andresag01 commented:
Only timing tests failed, so I think its OK

andresag01 commented:
@sbutcher-arm @hanno-arm: Could you please take another look? I will backport once the changes have been approved.

mpg commented:
@hanno-arm @sbutcher-arm Can you re-review this?

@andresag01 Can you go ahead and optimistically create backports, as an earlier version of this PR already had a near-approval from Hanno?

andresag01 commented:
@mpg: No problem, I will do that as soon as I can.

andresag01 commented:
@mpg: I have created/updated backports for Mbed TLS 2.16 and 2.7.

andresag01 commented on ChangeLog:
```diff
@@ -23,6 +29,8 @@ Features
      function to see for which parameter values it is defined. This feature is
      disabled by default. See its API documentation in config.h for additional
      steps you have to take when enabling it.
+   * Add a new function mbedtls_asn1_write_named_bitstring() to write ASN.1
```
Oh, I think it got put here when I rebased and I didnt notice it. I will move it up

andresag01 commented on include/mbedtls/x509_csr.h:
```diff
@@ -205,6 +205,14 @@ void mbedtls_x509write_csr_set_md_alg( mbedtls_x509write_csr *ctx, mbedtls_md_ty
+ *
+ * \note            The <code>decipherOnly</code> flag from the Key Usage
+ *                  extension is represented by bit 8 (i.e.
+ *                  <code>0x8000</code>), which cannot typically be represented
```
So, I think the confusion comes from the difference between in endianess in the computer and the encoding. Consider the bitfield we are trying to encode:
```
      KeyUsage ::= BIT STRING {
           digitalSignature        (0),
           nonRepudiation          (1), -- recent editions of X.509 have
                                -- renamed this bit to contentCommitment
           keyEncipherment         (2),
           dataEncipherment        (3),
           keyAgreement            (4),
           keyCertSign             (5),
           cRLSign                 (6),
           encipherOnly            (7),
           decipherOnly            (8) }
```
The way I read this is that `decipherOnly` is bit 8 (starting from 0). If you look at the value for `MBEDTLS_X509_KU_DECIPHER_ONLY` it clearly does not look like bit 8, but the code that writes the bitstring sort of "reverses" the position of the bits in each byte (see [here](https://github.com/ARMmbed/mbedtls/pull/2028/files#diff-9fb7e9933ee45102b152e2609eb047b3R351)). For example, our macro for `digitalSignature` that is supposed to be bit 0 is:
```
#define MBEDTLS_X509_KU_DIGITAL_SIGNATURE            (0x80)  /* bit 0 */
```

mpg replied:
Oh, that's absolutely right, now that you mention it I remember about this weird bit numbering. Thanks for reminding me!

hanno-arm commented on tests/suites/test_suite_asn1write.function:
```diff
@@ -128,3 +128,47 @@ void mbedtls_asn1_write_len( int len, data_t * asn1, int buf_len,
+
+    p = buf + GUARD_LEN + expected_asn1->len;
+
+    if ( is_named == 0 )
```
Minor: Space after `if`.
I suggest not to change this unless there's another blocking change, to not request re-approval.

hanno-arm commented on tests/suites/test_suite_asn1write.function:
```diff
@@ -128,3 +128,47 @@ void mbedtls_asn1_write_len( int len, data_t * asn1, int buf_len,
+        TEST_ASSERT( buf[GUARD_LEN + expected_asn1->len + i] == GUARD_VAL );
+    }
+
+    if ( result >= 0 )
```
Space after `if` here, too.

andresag01 commented:
@hanno-arm: Thanks for the review. I will not update the PR to fix the small style issue in the test code to avoid re-review...

andresag01 commented:
There are two approvals for this PR, so I removed the "needs: review" label.

mpg commented:
@andresag01 @hanno-arm I can't check the CI right now, could one of one check if the failure is a known issue and report your findings here? Thanks!

andresag01 commented:
@mpg: As far as I can tell, only timing tests failing.

mpg commented:
This has two approvals and a passing CI (except for known-fragile test), a ChangeLog entry and tests, so it would be ready for merge, but the backports need reviewing.

hanno-arm commented:
Ping @mpg on an opinion regarding https://github.com/ARMmbed/mbedtls/pull/2028/files#diff-030fafc49adaa9d23e735ab6a4cf816cR93; see the comment above.

mpg commented:
@hanno-arm Thanks for pinging me. I think this should not hold back this PR, as (1) it's pre-existing and (2) it doesn't affect observable behaviour in any way unless there is a bug elsewhere, so it's much less important than the observable compliance issue addressed by this PR, and we shouldn't hold back important fixes for trivial pre-existing issues.

mpg commented:
Backports are ready for merge, removing "needs: backports".

mpg commented:
This has two approvals and a (as good as) passing CI, a ChangeLog entry and tests, fixes a legitimate bug and has backports to all maintained branches ready, so it's ready for merge.

sbutcher-arm commented:
I've rebased the PR (to include the `crypto` submodule) and appended a commit to update the submodule to commit [`a78c958`](https://github.com/ARMmbed/mbed-crypto/commit/a78c958b17d75ddf63d8dd17255b6379dcbf259f) which includes the same PR merged into the `crypto` submodule.

@hanno-arm / @mpg - Can you please re-review this PR and approve?
